orm/serial_bucket.go

```diff
@@ -0,0 +1,384 @@
+	// return more than one value for the same key.
+	// All matching entities are appended to given destination slice. If no
+	// result was found, no error is returned and destination slice is not
+	// modified.
+	ByIndex(db weave.ReadOnlyKVStore, indexName string, key []byte, dest SerialModelSlicePtr) error
+
+	// IndexScan does a PrefixScan, but on the named index. This would let us eg. load all counters
+	// in order of their count. Or easily find the lowest or highest count.
+	IndexScan(db weave.ReadOnlyKVStore, indexName string, prefix []byte, reverse bool) (SerialModelIterator, error)
+
+	// Create saves given SerialModel in the database. Before inserting into
+	// database, SerialModel is validated using its Validate method.
+	// ID field must be unset so auto incremented ID is generated.
+	Create(db weave.KVStore, m SerialModel) error
+
+	// Upsert creates given SerialModel in the database or upserts of given SerialModel
+	// with given ID exists. ID field must be set.
+	Upsert(db weave.KVStore, m SerialModel) error
+
+	// Delete removes an entity with given primary key from the database.
+	// Returns ErrNotFound if an entity with given key does not exist.
+	Delete(db weave.KVStore, key []byte) error
+
+	// Has returns nil if an entity with given primary key value exists. It
+	// returns ErrNotFound if no entity can be found.
+	// Has is a cheap operation that that does not read the data and only
+	// checks the existence of it.
+	Has(db weave.KVStore, key []byte) error
+
+	// Register registers this buckets content to be accessible via query
+	// requests under the given name.
+	Register(name string, r weave.QueryRouter)
+}
+
+// NewSerialModelBucket returns a SerialModelBucket instance. This implementation relies on
+// a bucket instance. Final implementation should operate directly on the
+// KVStore instead.
+func NewSerialModelBucket(name string, m SerialModel, opts ...SerialModelBucketOption) SerialModelBucket {
+	b := NewBucket(name, m)
+
+	tp := reflect.TypeOf(m)
+	if tp.Kind() == reflect.Ptr {
+		tp = tp.Elem()
+	}
+
+	smb := &serialModelBucket{
+		b:          b,
+		idSeq:      b.Sequence("id"),
+		model:      tp,
+		bucketName: name,
+	}
+	for _, fn := range opts {
+		fn(smb)
+	}
+	return smb
+}
+
+// SerialModelBucketOption is implemented by any function that can configure
+// SerialModelBucket during creation.
+type SerialModelBucketOption func(smb *serialModelBucket)
+
+// indexInfo keeps information on index
+type indexInfo struct {
+	name string
+	// prefix is the kvstore prefix used for all items in the index
+	prefix []byte
+	unique bool
+}
+
+// WithIndexSerial configures the bucket to build an index with given name. All
+// entities stored in the bucket are indexed using value returned by the
+// indexer function. If an index is unique, there can be only one entity
+// referenced per index value.
+func WithIndexSerial(name string, indexer Indexer, unique bool) SerialModelBucketOption {
```

Comment:
I like these options and that they are only in the Build / New phase. But...

This can actually be called on a bucket after it returns from new.

Maybe add a `fixed bool` flag on `serialModelBucket` that is returned after applying the opts in `NewSerialModelBucket`, and then `WithIndexSerial` panics (or is a no op) if that flag is set.

Addresses the concern of Piotr and Alex of adding Indexes *after* using it.

Reply:
@husio @alpe ?

Reply:
This is possible, because `SerialModelBucketOption` takes a shortcut on what it accepts. "Usually" the argument would be a function that accepts a private configuration object, not the retruned by `NewXxx` function. For example:
```go
type options struct {                                                          
        index string                                                           
}                                                                              
                                                                               
type Option func(*options)                                                     
                                                                               
func WithIndexSerial(name string) Option {                                     
        return func(o *options) {          
                // No direct access to the bucket instance here. 
                // Only the options structure can be modified.                                    
                o.index = name                                                 
        }                                                                      
}                                                                              
                                                                               
func NewModel(args ...Option) *serialModelBucket {                             
        var o options                                                          
        for _, fn := range args {                                              
                fn(&o)                                                         
        }                                                                      
                                                                               
        var bucket serialModelBucket                                           
               
        // Inspect the options structure to decide which functionality was configured.                                                                
        if o.index {                                                           
                // stuff                                                       
                // bucket.index.xxx                                            
        }                                                                      
                                                                               
        return &bucket                                                         
}   
```

There is no `options` entity in this implementation. Instead, the bucket is modified directly. 

I am fine with both implementations - direct access and through options structure proxy. Implementation with the proxy is safer as it cannot change the model instance directly.